File: Python/utils/tree.py
from __future__ import annotations
from typing import Any

class TreeNode:
    def __init__(self, val: Any, left:TreeNode|None=None, right:TreeNode|None=None):
        self.val = val
        self.left = left
        self.right = right

    @classmethod
    def from_list(cls, nodes: list[Any]) -> TreeNode|None:
        if 0 == len(nodes):
            return None
        root: TreeNode = TreeNode(nodes[0])
        current_depth: list[TreeNode] = [root]
        next_depth: list[TreeNode] = []
        i: int = 1
        while len(nodes) > i:
            for node in current_depth:
                if len(nodes) > i and nodes[i] is not None:
                    node.left = TreeNode(nodes[i])
                    next_depth.append(node.left)
                if len(nodes) > i + 1 and nodes[i + 1] is not None:
                    node.right = TreeNode(nodes[i + 1])
                    next_depth.append(node.right)
                i += 2
            current_depth = next_depth
            next_depth = []
        return root

    def to_list(self) -> list[Any]:
        nodes: list[Any] = [self.val]
        current_depth: list[TreeNode] = [self]
        next_depth: list[TreeNode] = []
        while 0 != len(current_depth):
            for node in current_depth:
                if node.left is not None:
                    nodes.append(node.left.val)
                    next_depth.append(node.left)
                else:
                    nodes.append(None)
                if node.right is not None:
                    nodes.append(node.right.val)
                    next_depth.append(node.right)
                else:
                    nodes.append(None)
            current_depth = next_depth
            next_depth = []
        i: int = len(nodes) - 1
        while nodes[i] is None:
            i -= 1
        return nodes[0:i+1]

    def __eq__(self, other):
        # TODO: Implement a better way
        if self is None and other is None:
            return True
        if self is None or other is None:
            return False
        return self.to_list() == other.to_list()

    def __repr__(self):
        return str(self.to_list())

File: TypeScript/utils/tree.ts
export class TreeNode {
    val: any;
    left: TreeNode | null;
    right: TreeNode | null;

    constructor(val: any, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    static fromList(nodes: (any | null)[]): TreeNode | null {
        if (nodes.length === 0) return null;
        const root = new TreeNode(nodes[0]);
        let currentDepth: TreeNode[] = [root];
        let nextDepth: TreeNode[] = [];
        let i = 1;
        while (i < nodes.length) {
            for (const node of currentDepth) {
                if (i < nodes.length && nodes[i] !== null) {
                    node.left = new TreeNode(nodes[i]);
                    nextDepth.push(node.left);
                }
                if (i + 1 < nodes.length && nodes[i + 1] !== null) {
                    node.right = new TreeNode(nodes[i + 1]);
                    nextDepth.push(node.right);
                }
                i += 2;
            }
            currentDepth = nextDepth;
            nextDepth = [];
        }
        return root;
    }

    toList(): (any | null)[] {
        const nodes: (any | null)[] = [this.val];
        let currentDepth: TreeNode[] = [this];
        let nextDepth: TreeNode[] = [];
        while (currentDepth.length !== 0) {
            for (const node of currentDepth) {
                if (node.left !== null) {
                    nodes.push(node.left.val);
                    nextDepth.push(node.left);
                } else {
                    nodes.push(null);
                }
                if (node.right !== null) {
                    nodes.push(node.right.val);
                    nextDepth.push(node.right);
                } else {
                    nodes.push(null);
                }
            }
            currentDepth = nextDepth;
            nextDepth = [];
        }
        let end = nodes.length - 1;
        while (end >= 0 && nodes[end] === null) {
            end--;
        }
        return nodes.slice(0, end + 1);
    }

    toString(): string {
        return JSON.stringify(this.toList());
    }

    equals(other: TreeNode | null): boolean {
        if (other === null) return false;
        return JSON.stringify(this.toList()) === JSON.stringify(other.toList());
    }
}
